tempodb/blocklist/poller.go

```diff
@@ -226,69 +256,141 @@ func (p *Poller) pollTenantAndCreateIndex(ctx context.Context, tenantID string)
+		return nil, nil, err
 	}
 
-	bg := boundedwaitgroup.New(p.cfg.PollConcurrency)
-	chMeta := make(chan *backend.BlockMeta, len(blockIDs))
-	chCompactedMeta := make(chan *backend.CompactedBlockMeta, len(blockIDs))
+	readPreviousSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "readPreviousResults")
+	metas := previous.Metas(tenantID)
+	readPreviousSpan.SetTag("metas", len(metas))
+	compactedMetas := previous.CompactedMetas(tenantID)
+	readPreviousSpan.SetTag("compactedMetas", len(compactedMetas))
+	readPreviousSpan.Finish()
+
+	assembleSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "assemblePreviousResults")
+	mm := make(map[uuid.UUID]*backend.BlockMeta, len(metas))
+	cm := make(map[uuid.UUID]*backend.CompactedBlockMeta, len(compactedMetas))
+
+	for _, i := range metas {
+		mm[i.BlockID] = i
+	}
+
+	for _, i := range compactedMetas {
+		cm[i.BlockID] = i
+	}
+	assembleSpan.Finish()
+
+	mtx := sync.Mutex{}
+	newBlockList := make([]*backend.BlockMeta, 0, len(currentBlockIDs))
+	newCompactedBlocklist := make([]*backend.CompactedBlockMeta, 0, len(currentCompactedBlockIDs))
 	anyError := atomic.Error{}
 
-	for _, blockID := range blockIDs {
+	learnSpan, _ := opentracing.StartSpanFromContext(derivedCtx, "learnFromPrevious")
+	newBlockIDs := []uuid.UUID{}
+	for _, blockID := range currentBlockIDs {
+		// if we already have this block id in our previous list, use the existing data.
+		if v, ok := mm[blockID]; ok {
+			newBlockList = append(newBlockList, v)
+			continue
+		}
+		newBlockIDs = append(newBlockIDs, blockID)
+	}
+
+	for _, blockID := range currentCompactedBlockIDs {
+		// if we already have this block id in our previous list, use the existing data.
+		if v, ok := cm[blockID]; ok {
+			newCompactedBlocklist = append(newCompactedBlocklist, v)
+			continue
+		}
+
+		// TODO: Review this hack to avoid polling for compacted blocks that we
+		// know about.  But we need to poll this to ensure that we have the correct
+		// time. Probably.
+		// if v, ok := mm[blockID]; ok {
+		// 	chCompactedMeta <- &backend.CompactedBlockMeta{
+		// 		BlockMeta: *v,
+		// 		// CompactedTime: time.Now(),
+		// 	}
+		// 	continue
+		// }
+		newBlockIDs = append(newBlockIDs, blockID)
+	}
+	learnSpan.Finish()
+
+	pollSpan, pollSpanCtx := opentracing.StartSpanFromContext(derivedCtx, "pollUnknown")
+	pollSpan.SetTag("newBlockIDs", len(newBlockIDs))
+	bg := boundedwaitgroup.New(p.cfg.PollConcurrency)
+	for _, blockID := range newBlockIDs {
 		bg.Add(1)
-		go func(uuid uuid.UUID) {
+		go func(id uuid.UUID) {
 			defer bg.Done()
 
 			if p.cfg.PollJitterMs > 0 {
 				time.Sleep(time.Duration(rand.Intn(p.cfg.PollJitterMs)) * time.Millisecond)
 			}
 
-			m, cm, err := p.pollBlock(ctx, tenantID, uuid)
+			m, cm, pollBlockErr := p.pollBlock(pollSpanCtx, tenantID, id)
 			if m != nil {
-				chMeta <- m
-			} else if cm != nil {
-				chCompactedMeta <- cm
-			} else if err != nil {
+				mtx.Lock()
```

Comment:
lock and then defer an unlock over these detailed locks/unlocks

Reply:
This seems okay to me.  What's the reason for your suggestion?  Just readability?

Reply:
readability as well as reducing the likelihood of writing a deadlock

Reply:
Sorry, is there risk deadlock on the append?  What am I missing?

The only reason I'm pushing back here is that I'd rather not lock on error.  I can adjust to check error first and such, but locking only when appending.